src/util_cs/mod.rs

```diff
@@ -1,3 +1,86 @@
+use crate::LinearCombination;
+use ff::PrimeField;
+
 pub mod bench_cs;
 pub mod metric_cs;
 pub mod test_cs;
+
+pub type Constraint<Scalar> = (
+    LinearCombination<Scalar>,
+    LinearCombination<Scalar>,
+    LinearCombination<Scalar>,
+    String,
+);
+
+pub trait Comparable<Scalar: PrimeField> {
+    /// The `Comparable` trait allows comparison of two constraint systems which
+    /// implement the trait. The only non-trivial method, `delta`, has a default
+    /// implementation which supplies the desired behavior.
+    ///
+    /// Use `delta` to compare constraint systems. If they are not identical, the
+    /// returned `Delta` enum contains fine-grained information about how they
+    /// differ. This can be especially useful when debugging the situation in which
+    /// a constraint system is satisfied, but the corresponding Groth16 proof does
+    /// not verify.
+    ///
+    /// Example usage:
+    ///
+    /// ```norun
+    /// let delta = cs.delta(&cs_blank);
+    /// assert!(delta == Delta::Equal);
+    /// ```
+    fn num_inputs(&self) -> usize;
+    fn num_constraints(&self) -> usize;
+    fn inputs(&self) -> Vec<String>;
+    fn aux(&self) -> Vec<String>;
+    fn constraints(&self) -> &[Constraint<Scalar>];
+
+    fn delta<C: Comparable<Scalar>>(&self, other: &C) -> Delta<Scalar>
+    where
+        Scalar: PrimeField,
+    {
+        let input_count_matches = self.num_inputs() == other.num_inputs();
+        let constraint_count_matches = self.num_constraints() == other.num_constraints();
+
+        let inputs_match = self.inputs() == other.inputs();
+        let constraints_match = self.constraints() == other.constraints();
+
+        let equal =
+            input_count_matches && constraint_count_matches && inputs_match && constraints_match;
+
+        if !input_count_matches {
+            Delta::InputCountMismatch(self.num_inputs(), other.num_inputs())
+        } else if !constraint_count_matches {
+            Delta::ConstraintCountMismatch(self.num_constraints(), other.num_constraints())
+        } else if !constraints_match {
+            let c = self.constraints();
+            let o = other.constraints();
+
+            let mismatch = c
+                .iter()
+                .zip(o)
+                .filter(|(a, b)| a != b)
+                .enumerate()
```

Comment:
I believe this gives the wrong index, you need to enumerate begore filtering if you want the absplute number

Reply:
Good catch, fixed. The underlying problem this was used to diagnose was already fixed when I made this change, which is otherwise worthwhile for reasons I noted elsewhere.